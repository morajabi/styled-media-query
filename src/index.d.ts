// Type definitions for styled-media-query 2.1.1
// Project: https://github.com/morajabi/styled-media-query
// Definitions by: Fran√ßois Best <https://github.com/franky47>
// Requires @types/styled-components definitions ^4.1
// TypeScript version: 3.3.3

import {
  ThemedStyledProps,
  InterpolationValue,
  FlattenInterpolation
} from 'styled-components'

type InterpolationFunction<Props, Theme> = (
  props: ThemedStyledProps<Props, Theme>
) => InterpolationValue | FlattenInterpolation<ThemedStyledProps<Props, Theme>>

type GeneratorFunction<Props, Theme> = (
  strings: TemplateStringsArray,
  ...interpolations: (
    | InterpolationValue
    | InterpolationFunction<Props, Theme>
    | FlattenInterpolation<ThemedStyledProps<Props, Theme>>)[]
) => any

// --

export interface MediaGenerator<Breakpoints, Theme> {
  lessThan: <Props>(
    breakpoint: keyof Breakpoints
  ) => GeneratorFunction<Props, Theme>
  greaterThan: <Props>(
    breakpoint: keyof Breakpoints
  ) => GeneratorFunction<Props, Theme>
  between: <Props>(
    fist: keyof Breakpoints,
    second: keyof Breakpoints
  ) => GeneratorFunction<Props, Theme>
}

// --

export interface DefaultBreakpoints {
  huge: string
  large: string
  medium: string
  small: string
}

export const defaultBreakpoints: DefaultBreakpoints

// --

export function generateMedia<Breakpoints = DefaultBreakpoints, Theme = any>(
  breakpoints?: Breakpoints
): MediaGenerator<Breakpoints, Theme>

// --

declare const media: MediaGenerator<DefaultBreakpoints, any>

export default media

// Convertors --

export function pxToEm<B>(breakpoints: B, ratio?: number): B
export function pxToRem<B>(breakpoints: B, ratio?: number): B
