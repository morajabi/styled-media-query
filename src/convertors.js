/**
 * Converts breakpoint units in px to rem or em
 * @param {Object} breakpoints - an object containing breakpoint names as keys and the width as value
 * @param {number} ratio [16] - size of 1 rem in px. What is your main font-size in px?
 * @param {'rem' | 'em'} unit
 */
function pxToEmOrRem(breakpoints, ratio = 16, unit) {
  const newBreakpoints = {};

  for (let key in breakpoints) {
    const point = breakpoints[key];

    if (String(point).includes('px')) {
      newBreakpoints[key] = +(parseInt(point) / ratio) + unit;
      continue;
    }

    newBreakpoints[key] = point;
  }

  return newBreakpoints;
}

/**
 * Converts breakpoint units in px to em 
 * @param {Object} breakpoints - an object containing breakpoint names as keys and the width as value
 * @param {number} ratio [16] - size of 1em in px. What is your main font-size in px?
 */
export function pxToEm(breakpoints, ratio = 16) {
  return pxToEmOrRem(breakpoints, ratio, 'em');
}

/**
 * Converts breakpoint units in px to rem 
 * @param {Object} breakpoints - an object containing breakpoint names as keys and the width as value
 * @param {number} ratio [16] - size of 1rem in px. What is your main font-size in px?
 */
export function pxToRem(breakpoints, ratio = 16) {
  return pxToEmOrRem(breakpoints, ratio, 'rem');
}
