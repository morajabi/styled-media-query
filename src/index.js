import { css } from 'styled-components';
import { pxToEm, pxToRem } from './convertors';

export {
  pxToEm,
  pxToRem
};

/**
 * Default media breakpoints
 * @type {Object}
 */
export const defaultBreakpoints = {
  huge: '1440px',
  large: '1170px',
  medium: '768px',
  small: '450px',
};

function getSizeFromBreakpoint(breakpointValue, breakpoints = {}) {
  if (breakpoints[breakpointValue]) {
    return breakpoints[breakpointValue];
  } else if (parseInt(breakpointValue)) {
    return breakpointValue;
  } else {
    console.error('styled-media-query: No valid breakpoint or size specified for media.');
    return '0';
  }
}

/**
 * Media query generator
 * @param {Object} breakpoints - Map labels to breakpoint sizes
 * @return {Object} - Media generators for each breakpoint
 */
export function generateMedia(breakpoints = defaultBreakpoints) {
  const lessThan = (breakpoint) => (...args) => css`
    @media (max-width: ${getSizeFromBreakpoint(breakpoint, breakpoints)}) {
      ${css(...args)}
    }
  `;

  const greaterThan = (breakpoint) => (...args) => css`
    @media (min-width: ${getSizeFromBreakpoint(breakpoint, breakpoints)}) {
      ${css(...args)}
    }
  `;

  const between = (firstBreakpoint, secondBreakpoint) => (...args) => css`
    @media (min-width: ${getSizeFromBreakpoint(firstBreakpoint, breakpoints)}) and
      (max-width: ${getSizeFromBreakpoint(secondBreakpoint, breakpoints)}) {
      ${css(...args)}
    }
  `;

  const oldStyle = Object
    .keys(breakpoints)
    .reduce((acc, label) => {
      const size = breakpoints[label];

      acc.to[label] = (...args) => {
        console.warn(`styled-media-query: Use media.lessThan('${label}') instead of old media.to.${label} (Probably we'll deprecate it)`);
        return css`
          @media (max-width: ${size}) {
            ${css(...args)}
          }
        `;
      };

      acc.from[label] = (...args) => {
        console.warn(`styled-media-query: Use media.greaterThan('${label}') instead of old media.from.${label} (Probably we'll deprecate it)`);
        return css`
          @media (min-width: ${size}) {
            ${css(...args)}
          }
        `;
      };

      acc.styled[label] = (...args) => css(...args);

      return acc;
    },
    { to: {}, from: {}, styled: {} }
  );

  return Object.assign(
    {
      lessThan,
      greaterThan,
      between,
    },
    oldStyle,
  );
}

/**
 * Generate Media as a chained callable class by extending Function.
 * (Proposal for V3)
 */
export class GenerateMediaChained extends Function {
  /**
   * Constructor to chain `generateMedia()` and still bind the class to itself.
   * @param breakpoints   Object    The initial breakpoint object.
   * @return {any}
   */
  constructor(breakpoints = defaultBreakpoints) {
    // Call parent constuctor, making this a function.
    super('...args', 'return this.__self__.__call__(...args)');
    // Reestablish context by binding to this.
    const self = this.bind(this);
    this.__self__ =  self;
    // As we have to return an instance of `self` with bound this, create the
    // class variables on it. `callees` will later be a bound creator.
    self.callees = []
    self.media = generateMedia(breakpoints)
    return self;
  }

  /**
   * Wrapper around `lessThan()`.
   * At the moment we have to reset the length of `callees` as long as chaining
   * isn't discussed further.
   *
   * @param breakpoint    String|Number   Breakpoint to create Media-query from.
   * @return {GenerateMediaChained}
   */
  lessThan (breakpoint) {
    if (this.callees.length > 1) this.callees = []
    this.callees.push(this.media.lessThan(breakpoint))
    return this;
  }

  /**
   * Wrapper around `greaterThan()`
   *
   * @param breakpoint    String|Number   Breakpoint to create Media-query from.
   * @return {GenerateMediaChained}
   */
  greaterThan (breakpoint) {
    if (this.callees.length > 1) this.callees = []
    this.callees.push(this.media.greaterThan(breakpoint))
    return this;
  }

  /**
   * Wrapper around `between()`.
   *
   * @param firstBreakpoint   String|Number  Lower bound breakpoint.
   * @param secondBreakpoint  String|Number  Upper bound breakpoint.
   * @return {GenerateMediaChained}
   */
  between (firstBreakpoint, secondBreakpoint) {
    if (typeof firstBreakpoint === "string") this.css(firstBreakpoint)
    if (this.callees.length > 1) this.callees = []
    this.callees.push(this.media.between(firstBreakpoint, secondBreakpoint))
    return this;
  }

  /**
   * Finalizing function (like jQuerys' `end()`).
   * Tried to replicate the highlighting for styled-components with this, but
   * sadly doesn't work.
   *
   * @param args    String    The css to give to the functions in `callees`.
   * @return {*}
   */
  css (...args) {
    return this.callees[0](...args)
  }

  /**
   * Same as `css()`. But now our class is a callable.
   *
   * @param args    String    The css to give to the functions in `callees`.
   * @return {*}
   * @private
   */
  __call__ (...args) {
    return this.css(...args)
  }
}

/**
 * Media object with default breakpoints
 * @return {object} - Media generators for each size
 */
export default generateMedia();

/**
 * Usage: styled.div` ${media.from.medium`background: #000;`} `;
 * With this code, background for small and medium sizes will be `default` and for more than medium, will be `#000`
 */
