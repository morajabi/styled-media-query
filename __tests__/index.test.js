import 'jest-styled-components';
import { generateMedia, defaultBreakpoints } from '../src/index';
import React from 'react';
import renderer from 'react-test-renderer';
import styled from 'styled-components';
import { css } from 'styled-components';

const generatedDefaultMedia = generateMedia(defaultBreakpoints);

const Box = styled.div`
  background: black;

  ${generatedDefaultMedia.lessThan('medium')`
    background: red;
  `}

  ${generatedDefaultMedia.between('medium', 'large')`
    /* screen width is between 768px (medium) and 1170px (large) */
    background: green;
  `}

  ${generatedDefaultMedia.greaterThan('large')`
    /* screen width is greater than 1170px (large) */
    background: blue;
  `}
  
  ${generatedDefaultMedia.to.large`
    background: white;
  `}
  
  ${generatedDefaultMedia.from.small`
    background: yellow;
  `}
`;

describe(`generatedDefaultMedia()`, () => {
  it('expect <Box /> to have matching media style rules', () => {
    let tree = renderer.create(<Box/>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('background', 'red', {
      media: '(max-width:768px)',
    });
    expect(tree).toHaveStyleRule('background', 'green', {
      media: '(min-width:768px) and (max-width:1170px)',
    });
    expect(tree).toHaveStyleRule('background', 'blue', {
      media: '(min-width:1170px)',
    });
    expect(tree).toHaveStyleRule('background', 'white', {
      media: '(max-width:1170px)',
    });
    expect(tree).toHaveStyleRule('background', 'yellow', {
      media: '(min-width:450px)',
    });
  });
});

describe(`lessThan()`, () => {
  it('expect lessThan("medium") to create Array', () => {
    let tree = renderer
        .create(
            css`
        ${generatedDefaultMedia.lessThan('medium')`
    background: red
  `}
      `
        )
        .toJSON();
    let itArray = [
      '\n    @media (max-width: ',
      '768px',
      ') {\n      ',
      '\n    background: red\n  ',
      '\n    }\n  ',
    ];
    expect(tree).toEqual(expect.arrayContaining(itArray));
  });
});

describe(`between()`, () => {
  it('expect between("medium", "large") to create Array', () => {
    let tree = renderer
        .create(
            css`
        ${generatedDefaultMedia.between('medium', 'large')`
    background: green;
  `}
      `
        )
        .toJSON();
    let itArray = [
      '\n    @media (min-width: ',
      '768px',
      ') and\n      (max-width: ',
      '1170px',
      ') {\n      ',
      '\n    background: green;\n  ',
      '\n    }\n  ',
    ];
    expect(tree).toEqual(expect.arrayContaining(itArray));
  });
});

describe(`greaterThan()`, () => {
  it('expect greaterThan("large") to create Array', () => {
    let tree = renderer
        .create(
            css`
        ${generatedDefaultMedia.greaterThan('large')`
    background: blue;
  `}
      `
        )
        .toJSON();
    let itArray = [
      '\n    @media (min-width: ',
      '1170px',
      ') {\n      ',
      '\n    background: blue;\n  ',
      '\n    }\n  ',
    ];
    expect(tree).toEqual(expect.arrayContaining(itArray));
  });

  it('expect greaterThan(320) to create Array', () => {
    let tree = renderer
        .create(
            css`
        ${generatedDefaultMedia.greaterThan('320')`
    background: blue;
  `}
      `
        )
        .toJSON();
    let itArray = [
      '\n    @media (min-width: ',
      '320',
      ') {\n      ',
      '\n    background: blue;\n  ',
      '\n    }\n  ',
    ];
    expect(tree).toEqual(expect.arrayContaining(itArray));
  });

  it('expect greaterThan() to create Array', () => {
    let tree = renderer
        .create(
            css`
        ${generatedDefaultMedia.greaterThan('')`
    background: blue;
  `}
      `
        )
        .toJSON();
    let itArray = [
      '\n    @media (min-width: ',
      '0',
      ') {\n      ',
      '\n    background: blue;\n  ',
      '\n    }\n  ',
    ];
    expect(tree).toEqual(expect.arrayContaining(itArray));
  });
});

describe(`defaultBreakpoints`, () => {
  it('expect defaultBreakpoints to equal sizes', () => {
    expect(defaultBreakpoints.small).toEqual('450px');
    expect(defaultBreakpoints.medium).toEqual('768px');
    expect(defaultBreakpoints.large).toEqual('1170px');
    expect(defaultBreakpoints.huge).toEqual('1440px');
  });
});
