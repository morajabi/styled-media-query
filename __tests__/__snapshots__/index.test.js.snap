// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect <Box /> to have matching media style rules 1`] = `
.c0 {
  background: black;
}

@media (max-width:768px) {
  .c0 {
    background: red;
  }
}

@media (min-width:768px) and (max-width:1170px) {
  .c0 {
    background: green;
  }
}

@media (min-width:1170px) {
  .c0 {
    background: blue;
  }
}

@media (max-width:1170px) {
  .c0 {
    background: white;
  }
}

@media (min-width:450px) {
  .c0 {
    background: yellow;
  }
}

<div
  className="c0"
/>
`;
